File: backend-midway/src/service/error.service.ts
import { Provide, Inject, Init } from '@midwayjs/core';
import { getModelForClass, ReturnModelType } from '@typegoose/typegoose';
import { ErrorLog } from '../entity/error.entity';
import { InfluxDBService } from './influxdb.service';
import { CacheService } from './cache.service';
import * as crypto from 'crypto';

@Provide()
export class ErrorService {
  errorModel: ReturnModelType<typeof ErrorLog>;

  @Init()
  async init() {
    this.errorModel = getModelForClass(ErrorLog);
  }

  @Inject()
  influxDBService: InfluxDBService;

  @Inject()
  cacheService: CacheService;

  /**
   * 生成错误指纹
   */
  private generateErrorHash(message: string, stack?: string, url?: string): string {
    const content = `${message}|${stack?.substring(0, 200) || ''}|${url || ''}`;
    return crypto.createHash('md5').update(content).digest('hex');
  }

  /**
   * 上报错误
   */
  async report(data: {
    projectId: string;
    type: string;
    message: string;
    stack?: string;
    url?: string;
    line?: number;
    col?: number;
    userAgent?: string;
    userId?: string;
    sessionId?: string;
  }): Promise<ErrorLog> {
    const timestamp = new Date();
    const errorHash = this.generateErrorHash(data.message, data.stack, data.url);

    // 尝试查找已存在的相同错误（聚合）
    // 使用 lean() 避免类型转换问题，特别是对于可能类型不一致的旧数据
    const existingError = await this.errorModel.findOne({
      projectId: data.projectId,
      errorHash,
    }).lean();

    let error: ErrorLog;

    if (existingError) {
      // 更新已有错误的计数和最后出现时间
      const currentUser = data.userId || data.sessionId || 'anonymous';
      const errorId = existingError._id!.toString();
      
      // 使用 Redis Set 来精确统计受影响用户数
      const userSetKey = `error:users:${errorId}`;
      const isNewUser = await this.cacheService.addToUserSet(userSetKey, currentUser);
      
      // 获取实际的受影响用户数（从 Redis Set）
      const actualUserCount = await this.cacheService.getUserSetSize(userSetKey);
      
      // 确保 affectedUsers 是数字类型
      const affectedUsersCount = Number(actualUserCount) || 1;
      
      // 更新错误记录
      const updateData: any = {
        $inc: { count: 1 },
        $set: { 
          lastSeen: timestamp,
          affectedUsers: affectedUsersCount, // 确保是数字类型
        },
      };
      
      error = await this.errorModel.findByIdAndUpdate(
        existingError._id,
        updateData,
        { new: true }
      );
    } else {
      // 创建新错误记录
      // 明确指定字段，避免 ...data 展开时包含不应该的字段
      error = await this.errorModel.create({
        projectId: data.projectId,
        type: data.type,
        message: data.message,
        stack: data.stack,
        url: data.url,
        line: data.line,
        col: data.col,
        userAgent: data.userAgent,
        userId: data.userId,
        sessionId: data.sessionId,
        timestamp,
        errorHash,
        count: 1,
        firstSeen: timestamp,
        lastSeen: timestamp,
        affectedUsers: 1,
      });
      
      // 初始化用户集合
      const currentUser = data.userId || data.sessionId || 'anonymous';
      const errorId = error._id!.toString();
      const userSetKey = `error:users:${errorId}`;
      await this.cacheService.addToUserSet(userSetKey, currentUser);
    }

    // 写入 InfluxDB（用于时序统计）
    await this.influxDBService.writeErrorCount(data.projectId, data.type);

    // 更新实时计数
    await this.cacheService.incrTodayErrorCount(data.projectId);
    
    // 更新今日指定类型的错误计数（只统计js和promise）
    if (data.type === 'js' || data.type === 'promise') {
      await this.cacheService.incrTodayErrorCountByType(data.projectId, data.type);
    }

    // 清除统计缓存
    await this.cacheService.del(`error:stats:${data.projectId}`);
    await this.cacheService.del(`dashboard:${data.projectId}`);

    return error;
  }

  /**
   * 获取错误列表
   */
  async findList(params: {
    projectId: string;
    type?: string;
    page?: number;
    pageSize?: number;
    startTime?: string;
    endTime?: string;
  }): Promise<{ data: ErrorLog[]; total: number }> {
    const { projectId, type, page = 1, pageSize = 20, startTime, endTime } = params;

    const query: any = { projectId };

    if (type) {
      query.type = type;
    }

    if (startTime || endTime) {
      query.timestamp = {};
      if (startTime) query.timestamp.$gte = new Date(startTime);
      if (endTime) query.timestamp.$lte = new Date(endTime);
    }

    const [data, total] = await Promise.all([
      this.errorModel
        .find(query)
        .sort({ lastSeen: -1 })
        .skip((page - 1) * pageSize)
        .limit(pageSize),
      this.errorModel.countDocuments(query),
    ]);

    return { data, total };
  }

  /**
   * 获取错误详情
   */
  async findById(id: string): Promise<ErrorLog | null> {
    return this.errorModel.findById(id);
  }

  /**
   * 获取错误统计
   */
  async getStats(projectId: string): Promise<any> {
    // 尝试从缓存获取
    const cached = await this.cacheService.getErrorStats(projectId);
    if (cached) {
      return cached;
    }

    // 检查数据库连接
    if (!this.errorModel) {
      return { total: 0, typeStats: {}, timeStats: {} };
    }

    try {
      // 获取总错误数 - 使用聚合统计，累加所有错误的count字段
      // 因为错误是聚合的，count字段表示该错误发生的次数
      const totalResult = await this.errorModel.aggregate([
        { $match: { projectId } },
        { $group: { _id: null, total: { $sum: '$count' } } },
      ]);
      const total = totalResult.length > 0 ? totalResult[0].total : 0;

      // 按类型统计
      const typeStats = await this.errorModel.aggregate([
        { $match: { projectId } },
        { $group: { _id: '$type', count: { $sum: '$count' } } },
      ]);

      const typeStatsMap: Record<string, number> = {};
      typeStats.forEach(item => {
        typeStatsMap[item._id] = item.count;
      });

      // 获取时间趋势
      const trend = await this.influxDBService.queryErrorTrend(projectId, 7);
      const timeStats: Record<string, number> = {};
      trend.forEach((item: any) => {
        const date = new Date(item.time).toISOString().split('T')[0];
        timeStats[date] = item.count || 0;
      });

      const stats = {
        total,
        typeStats: typeStatsMap,
        timeStats,
      };

      // 缓存结果
      await this.cacheService.setErrorStats(projectId, stats);

      return stats;
    } catch (error) {
      console.warn('Error stats query failed:', error.message);
      return { total: 0, typeStats: {}, timeStats: {} };
    }
  }

  /**
   * 获取今日错误类型统计（只统计js和promise）
   * 使用 Redis 实时计数，更准确
   */
  async getTodayTypeStats(projectId: string): Promise<Record<string, number>> {
    try {
      // 从 Redis 获取今日的 js 和 promise 错误计数
      const [jsCount, promiseCount] = await Promise.all([
        this.cacheService.getTodayErrorCountByType(projectId, 'js'),
        this.cacheService.getTodayErrorCountByType(projectId, 'promise'),
      ]);

      const typeStatsMap: Record<string, number> = {};
      if (jsCount > 0) {
        typeStatsMap['js'] = jsCount;
      }
      if (promiseCount > 0) {
        typeStatsMap['promise'] = promiseCount;
      }

      // 调试日志
      console.log('Today type stats from Redis:', {
        projectId,
        jsCount,
        promiseCount,
        typeStatsMap
      });

      return typeStatsMap;
    } catch (error) {
      console.warn('Get today type stats failed:', error.message);
      return {};
    }
  }

  /**
   * 删除超过指定天数的错误数据
   */
  async deleteOldData(projectId: string, days: number = 30): Promise<void> {
    if (!this.errorModel) return;

    try {
      const cutoffDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000);
      const result = await this.errorModel.deleteMany({
        projectId,
        createdAt: { $lt: cutoffDate },
      });
      console.log(`✅ Deleted ${result.deletedCount} error records older than ${days} days for project: ${projectId}`);
    } catch (error: any) {
      console.error('Failed to delete old error data:', error.message);
      // 不抛出错误，避免影响其他数据源的清理
    }
  }
}

File: backend-springboot/src/main/java/com/monitor/service/ErrorService.java
package com.monitor.service;

import com.monitor.entity.ErrorLog;
import com.monitor.repository.ErrorLogRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.security.MessageDigest;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * 错误服务
 */
@Service
@RequiredArgsConstructor
public class ErrorService {
    
    private final ErrorLogRepository errorLogRepository;
    private final InfluxDBService influxDBService;
    private final CacheService cacheService;
    
    /**
     * 生成错误指纹
     */
    private String generateErrorHash(String message, String stack, String url) {
        try {
            String content = message + "|" + 
                (stack != null ? stack.substring(0, Math.min(200, stack.length())) : "") + "|" + 
                (url != null ? url : "");
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hashBytes = md.digest(content.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hashBytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (Exception e) {
            String content = message + "|" + 
                (stack != null ? stack.substring(0, Math.min(200, stack.length())) : "") + "|" + 
                (url != null ? url : "");
            return String.valueOf(content.hashCode());
        }
    }
    
    /**
     * 上报错误
     */
    public ErrorLog report(String projectId, String type, String message, String stack,
                          String url, Integer line, Integer col, String userAgent,
                          String userId, String sessionId) {
        Date timestamp = new Date();
        String errorHash = generateErrorHash(message, stack, url);
        
        // 查找已存在的相同错误
        Optional<ErrorLog> existingErrorOpt = errorLogRepository.findByProjectIdAndErrorHash(projectId, errorHash);
        
        ErrorLog error;
        if (existingErrorOpt.isPresent()) {
            // 更新已有错误
            ErrorLog existingError = existingErrorOpt.get();
            String currentUser = userId != null ? userId : (sessionId != null ? sessionId : "anonymous");
            String errorId = existingError.getId();
            
            // 使用 Redis Set 统计受影响用户数
            cacheService.addToUserSet(errorId, currentUser);
            Long actualUserCount = cacheService.getUserSetSize(errorId);
            
            existingError.setCount(existingError.getCount() + 1);
            existingError.setLastSeen(timestamp);
            existingError.setAffectedUsers(actualUserCount.intValue());
            
            error = errorLogRepository.save(existingError);
        } else {
            // 创建新错误记录
            error = new ErrorLog();
            error.setProjectId(projectId);
            error.setType(type != null ? type : "js");
            error.setMessage(message);
            error.setStack(stack);
            error.setUrl(url);
            error.setLine(line);
            error.setCol(col);
            error.setUserAgent(userAgent);
            error.setUserId(userId);
            error.setSessionId(sessionId);
            error.setTimestamp(timestamp);
            error.setErrorHash(errorHash);
            error.setCount(1);
            error.setFirstSeen(timestamp);
            error.setLastSeen(timestamp);
            error.setAffectedUsers(1);
            
            error = errorLogRepository.save(error);
            
            // 初始化用户集合
            String currentUser = userId != null ? userId : (sessionId != null ? sessionId : "anonymous");
            cacheService.addToUserSet(error.getId(), currentUser);
        }
        
        // 写入 InfluxDB
        influxDBService.writeErrorCount(projectId, type != null ? type : "js");
        
        // 更新实时计数
        cacheService.incrTodayErrorCount(projectId);
        
        // 清除统计缓存
        cacheService.delete("error:stats:" + projectId);
        cacheService.delete("dashboard:" + projectId);
        
        return error;
    }
    
    /**
     * 获取错误列表
     */
    public Map<String, Object> findList(String projectId, String type, Integer page, Integer pageSize,
                                        String startTime, String endTime) {
        Pageable pageable = PageRequest.of(
            (page != null && page > 0 ? page - 1 : 0),
            (pageSize != null && pageSize > 0 ? pageSize : 20),
            Sort.by(Sort.Direction.DESC, "lastSeen")
        );
        
        Page<ErrorLog> pageResult;
        if (type != null && !type.isEmpty()) {
            pageResult = errorLogRepository.findByProjectIdAndType(projectId, type, pageable);
        } else {
            pageResult = errorLogRepository.findByProjectId(projectId, pageable);
        }
        
        Map<String, Object> result = new HashMap<>();
        result.put("data", pageResult.getContent());
        result.put("total", pageResult.getTotalElements());
        result.put("page", page != null && page > 0 ? page : 1);
        result.put("pageSize", pageSize != null && pageSize > 0 ? pageSize : 20);
        
        return result;
    }
    
    /**
     * 获取错误详情
     */
    public Optional<ErrorLog> findById(String id) {
        return errorLogRepository.findById(id);
    }
    
    /**
     * 获取错误统计
     */
    public Map<String, Object> getStats(String projectId) {
        // 尝试从缓存获取
        Map<String, Object> cached = cacheService.get("error:stats:" + projectId, Map.class);
        if (cached != null) {
            return cached;
        }
        
        try {
            // 获取总错误数
            long total = errorLogRepository.countByProjectId(projectId);
            
            // 按类型统计
            List<ErrorLog> typeStats = errorLogRepository.findTypeStatsByProjectId(projectId);
            Map<String, Long> typeStatsMap = new HashMap<>();
            for (ErrorLog error : typeStats) {
                typeStatsMap.put(error.getType(), 
                    typeStatsMap.getOrDefault(error.getType(), 0L) + error.getCount());
            }
            
            Map<String, Object> stats = new HashMap<>();
            stats.put("total", total);
            stats.put("typeStats", typeStatsMap);
            stats.put("timeStats", new HashMap<>()); // TODO: 从 InfluxDB 查询时间趋势
            
            // 缓存结果
            cacheService.set("error:stats:" + projectId, stats, 60L);
            
            return stats;
        } catch (Exception e) {
            System.err.println("Error stats query failed: " + e.getMessage());
            Map<String, Object> stats = new HashMap<>();
            stats.put("total", 0L);
            stats.put("typeStats", new HashMap<>());
            stats.put("timeStats", new HashMap<>());
            return stats;
        }
    }
}

